jumanji/environments/logic/sudoku/env.py

```diff
@@ -0,0 +1,207 @@
+    """A JAX implementation of the sudoku game.
+
+    - observation: `Observation`
+        - board: jax array (int32) of shape (9,9):
+            empty cells are represented by -1, and filled cells are represented by 0-8.
+        - action_mask: jax array (bool) of shape (9,9,9):
+            indicates which actions are valid.
+
+    - action:
+        multi discrete array containing the square to write a digit, and the digits
+        to input.
+
+    - reward: jax array (float32):
+        1 at the end of the episode if the board is valid
+        0 otherwise
+
+
+    - state: `State`
+        - board: jax array (int32) of shape (9,9):
+            empty cells are represented by -1, and filled cells are represented by 0-8.
+
+        - action_mask: jax array (bool) of shape (9,9,9):
+            indicates which actions are valid (empty cells and valid digits).
+
+        - key: jax array (int32) of shape (2,) used for seeding initial sudoku
+            configuration.
+
+    ```python
+    from jumanji.environments import Sudoku
+    env = Sudoku()
+    key = jax.random.key(0)
+    state, timestep = jax.jit(env.reset)(key)
+    env.render(state)
+    action = env.action_spec().generate_value()
+    state, timestep = jax.jit(env.step)(state, action)
+    env.render(state)
+    ```
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        if generator is None:
+            file_path = os.path.dirname(os.path.abspath(__file__))
+            database_file = DATABASES["mixed"]
+            database = jnp.load(os.path.join(file_path, "data", database_file))
+
+        self._generator = generator or DatabaseGenerator(database=database)
+        self._reward_fn = reward_fn or SparseRewardFn()
+        self._viewer = viewer or SudokuViewer()
+
+    def __repr__(self) -> str:
+        return f"Sudoku(grid_size={BOARD_WIDTH}x{BOARD_WIDTH})"
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        state = self._generator(key)
+        obs = Observation(board=state.board, action_mask=state.action_mask)
+        timestep = restart(observation=obs)
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        # check if action is valid
+        invalid = ~state.action_mask[tuple(action)]
+        updated_board = apply_action(action=action, board=state.board)
+        updated_action_mask = get_action_mask(board=updated_board)
+
+        winning = is_puzzle_solved(updated_board)
+
+        # creating next state
+        next_state = State(
+            board=updated_board, action_mask=updated_action_mask, key=state.key
+        )
+
+        no_actions_avalaible = ~jnp.any(updated_action_mask)
+
+        # computing terminal condition
+        done = invalid | winning | no_actions_avalaible
```

Comment:
Is `winning` not included in `no_actions_avalaible` already?

Reply:
I think it is, removing it